tests-integration/poller_test.go

```diff
@@ -461,3 +463,75 @@ func TestPollerExpiryEnsurePollingRace(t *testing.T) {
 		t.Fatalf("Should have got 401 http response; got %d\n%s", status, resBytes)
 	}
 }
+
+// Regression test for the bugfix for https://github.com/matrix-org/sliding-sync/issues/287#issuecomment-1706522718
+// Specifically, we could cache the failure and never tell the poller about new tokens, wedging the client(!). This
+// seems to have been due to the following:
+//   - client hits sync for the first time. We /whoami and remember the token->user mapping in TokensTable.
+//   - client syncing + poller syncing, everything happy.
+//   - token expires. OnExpiredToken is sent to EnsurePoller which removes the entry from EnsurePoller and nukes the conns.
+//   - client hits sync, gets 400 M_UNKNOWN_POS due to nuked conns.
+//   - client hits a fresh /sync: for whatever reason, the token is NOT 401d there and then by the /whoami lookup failing.
+//     Maybe failed to remove the token, but don't see any logs to suggest this. Seems to be an OIDC thing.
+//   - EnsurePoller starts a poller, which immediately 401s as the token is expired.
+//   - OnExpiredToken is sent first, which removes the entry in EnsurePoller.
+//   - OnInitialSyncComplete[success=false] is sent after, which MAKES A NEW ENTRY with success=false.
+//   - proxy sends back 401 M_UNKNOWN_TOKEN.
+//   - At this point the proxy is wedged. Any token, no matter how valid they are, will not hit EnsurePoller because
+//     we cached success=false for that (user,device).
+//
+// Traceable in the logs which show spam of this log line without "Poller: v2 poll loop started" interleaved.
+//
+//	12:45:33 ERR EnsurePolling failed, returning 401 conn=encryption device=xx user=@xx:xx.xx
+//
+// To test this failure mode we:
+// - Create Alice and sync her poller.
+// - Expire her token immediately, just like the test TestPollerExpiryEnsurePollingRace
+// - Do another request with a valid new token, this should succeed.
+func TestPollerExpiryEnsurePollingRaceDoesntWedge(t *testing.T) {
+	newToken := "NEW_ALICE_TOKEN"
+	pqString := testutils.PrepareDBConnectionString()
+	v2 := runTestV2Server(t)
+	defer v2.close()
+	v3 := runTestServer(t, v2, pqString)
+	defer v3.close()
+
+	v2.addAccount(t, alice, aliceToken)
+
+	// Arrange the following:
+	// 1. A request arrives from an unknown token.
+	// 2. The API makes a /whoami lookup for the new token. That returns without error.
+	// 3. The old token expires.
+	// 4. The poller tries to call /sync but finds that the token has expired.
+	// NEW 5. Using a "new token" works.
```

Comment:
Does this make the previous test I wrote above redundant? I.e. is is the same stuff + new, or is there some difference in the initial setup/assertions too?

Reply:
It's the same stuff + NEW, which I copy/pasta'd. Having 2 is still useful insofar that if your one passes but this new one fails, we know it's due to the extra step.